lib/spack/spack/version.py

```diff
@@ -335,387 +366,198 @@ def up_to(self, index):
-            _, self.ref_version = self._generate_separators_and_components(self.ref_version_str)
-            self.user_supplied_reference = True
-        else:
-            self.ref = pruned_string
+    __slots__ = ["ref", "has_git_prefix", "is_commit", "_ref_lookup", "_ref_version"]
 
-        self.is_commit = bool(len(self.ref) == 40 and COMMIT_VERSION.match(self.ref))
-        self.is_ref = git_prefix  # is_ref False only for comparing to VersionBase
-        self.is_ref |= bool(self.is_commit)
+    def __init__(self, string: str):
+        # An object that can lookup git refs to compare them to versions
+        self._ref_lookup: Optional[CommitLookup] = None
 
-        # ensure git.<hash> and <hash> are treated the same by dropping 'git.'
-        # unless we are assigning a version with =
-        canonical_string = self.ref if (self.is_commit and not self.ref_version) else string
-        super(GitVersion, self).__init__(canonical_string)
+        # This is the effective version.
+        self._ref_version: Optional[StandardVersion]
 
-    def _cmp(self, other_lookups=None):
-        # No need to rely on git comparisons for develop-like refs
-        if len(self.version) == 2 and self.isdevelop():
-            return self.version
+        self.has_git_prefix = string.startswith("git.")
 
-        # If we've already looked this version up, return cached value
-        if self.ref_version is not None:
-            return self.ref_version
+        # Drop `git.` prefix
+        normalized_string = string[4:] if self.has_git_prefix else string
 
-        ref_lookup = self.ref_lookup or other_lookups
+        if "=" in normalized_string:
+            # Store the git reference, and parse the user provided version.
+            self.ref, spack_version = normalized_string.split("=")
+            self._ref_version = StandardVersion(
+                spack_version, *parse_string_components(spack_version)
+            )
+        else:
+            # The ref_version is lazily attached after parsing, since we don't know what
+            # package it applies to here.
+            self._ref_version = None
+            self.ref = normalized_string
 
-        if self.is_ref and ref_lookup:
-            ref_info = ref_lookup.get(self.ref)
-            if ref_info:
-                prev_version, distance = ref_info
+        # Used by fetcher
+        self.is_commit: bool = len(self.ref) == 40 and bool(COMMIT_VERSION.match(self.ref))
 
-                if prev_version is None:
-                    prev_version = "0"
+    @property
+    def ref_version(self) -> StandardVersion:
+        # Return cached version if we have it
+        if self._ref_version is not None:
+            return self._ref_version
+
+        if self.ref_lookup is None:
+            raise RuntimeError(
+                f"git ref '{self.ref}' cannot be looked up: "
+                "call attach_git_lookup_from_package first"
+            )
```

Comment:
This should probably be an assertion rather than a runtime error, but see my note above on existing installations with unresolved git versions.  We need some type of solution for that other than "reinstall everything"

Reply:
I made it into a more specific error so it can be caught, which is necessary in some places